### ISSUE TITLE
Add more calculation methods

### ISSUE THREAD

**Issue by TorCzaf** [meypod/al-azan#66]
Hi , can you please add more calculation methods like:
-Ministry of religious affairs - Tunisia
-Ministry of Islamic affairs - Morocco
-Ministry of religious affairs and wakfs - Algeria 
-London unified prayer times
-Jabatan kemajuan Islam - Malaysia
-Mosquee de Paris  - France
-Angle 18° France
-Angle 15° France

And maybe there is more.

====

And also it would be good to add a button to share the app via Bluetooth in case if the sender and the receiver both have no WiFi/internet.
I took this idea from F-droid store itself.
This could be useful for people who travel in desert or in unknown places or be in unpopular areas...ETC

**Comment by meypod** [comment]
Hi
>-Ministry of religious affairs - Tunisia
-Ministry of Islamic affairs - Morocco
-Ministry of religious affairs and wakfs - Algeria
-London unified prayer times
-Jabatan kemajuan Islam - Malaysia
-Mosquee de Paris - France
-Angle 18° France
-Angle 15° France

Can you provide link to the official page of these methods ? so I can take their exact parameters ?
We can also add a "Custom" method which will allow users to enter their own angles 

> And also it would be good to add a button to share the app via Bluetooth in case if the sender and the receiver both have no WiFi/internet.

Some android manufactures already have this feature built-in to the OS
some of file managers do as well
if you have the file, all new android OS (Android 6+) also have the "Nearby Share" feature which allows you to send files without internet using WiFi or bluetooth to any device near you that also has the feature.

**Comment by meypod** [comment]
> You mean APK?

yeah I meant APK

I prefer to let third party apps handle it. I'm pretty sure there's a lot of apps you can use to get the APK file. check this app for example, I just found it: https://f-droid.org/en/packages/be.brunoparmentier.apkshare/
it works just as you want it to, plus you can also share other apps as well :)

**Comment by AI-IMED** [comment]
Hi @meypod  Many Twelver Muslims in various countries of the world, as well as many applications, rely on (Astronomical Research Center) to know the prayer times


I wonder, can you add this method of calculation to this great application?

https://english.nojumi.org/prayertimes

**Comment by meypod** [comment]
Hi @AI-IMED ,
The link you have given does not mention any angles that is used for calculation of the prayer times
since the app calculates everything offline, we need the angles and methods used by them

have you tried the "Shia Ithna Ashari, Leva Institute, Qum" method to see how it compares ?

**Comment by AI-IMED** [comment]
> The link you have given does not mention any angles that is used for calculation of the prayer times
since the app calculates everything offline, we need the angles and methods used by them

It saddened me , I depend on it because it gives very accurate prayer times.

> have you tried the "Shia Ithna Ashari, Leva Institute, Qum" or "Shia, Institute of Geophysics, University of Tehran" method to see how it compares ?

Yes, I tested that, it is not accurate, I have to adjust the prayer times, but even with adjusting the times, after two months or more, it makes a difference in the prayer times again

**Comment by meypod** [comment]
I'm sorry but without knowing the exact details of their calculation there's not much I can do

> Yes, I tested that, it is not accurate, I have to adjust the prayer times, but even with adjusting the times, after two months or more, it makes a difference in the prayer times again

Make sure to try changing the advanced settings in the calculation settings, if the calculated time is close, one of those may help adjust your time correctly

**Comment by AI-IMED** [comment]
I forgot to mention, It is necessary to add midnight to the calculation adjustment options, because there is a difference of opinion among the jurists of the Twelver doctrine, some of whom believe that the time of midnight is between sunset and dawn, and some of them believe that it is between sunset and sunrise

**Comment by meypod** [comment]
I use https://github.com/batoulapps/adhan-js library, I wanted to add adjustments for the midnight time, but it is provided by the library, and it doesn't provide a way to adjust it

please create an issue there, if it gets added there, I can add it as well

**Comment by meypod** [comment]
> Yes, I tested that, it is not accurate, I have to adjust the prayer times, but even with adjusting the times, after two months or more, it makes a difference in the prayer times again

"Shia, Institute of Geophysics, University of Tehran" method seems to match the times provided by the website you mentioned with only around 1 minute of difference which is usually due to rounding

**Comment by AI-IMED** [comment]
Yes, indeed, the difference is 1 minute, this calculation method seems better.

**Comment by meypod** [comment]
Well that would make it a lot easier, they got all the parameters [it seems](https://github.com/Five-Prayers/five-prayers-android/blob/9eda93b38dc226232b19ddaa3292d15ea311b6fb/app/src/main/java/com/hbouzidi/fiveprayers/timings/calculations/CalculationMethodEnum.java#L24)
the only issue is how credible they are (have you tried comparing times ?)

for example https://www.londonprayertimes.com/ takes time from https://www.eastlondonmosque.org.uk/prayer-times-and-calendar-explained which are based on time from http://astro.ukho.gov.uk/psp/index_beta.html which I can't find the calculation parameters from

**Comment by meypod** [comment]
> Also Speaking of calculations, can you also make the app to do online calculation and not only offline calculation if possible?

Not really, I try to keep app offline as much as possible

compare times with the official links with the times of the same method in five prayers app

**Comment by meypod** [comment]
> As you command :)

sorry, I didn't mean it like that

> At least not that kind of errors that may affect app installation am I right? Or you mean errors that may affect offline mode and everything in the app?

online, means it requires a network connection, and a backend
many things can go wrong, handling states between disconnections adds some complexity and unreliability, API may go down, it may become paid, etc

though one of the reasons is that in future if I manage to do something about Qibla map feature, I may remove internet permission from app to make it completely offline, which is better for privacy

There are other apps with mosques time like https://github.com/mawaqit/android-tv-app , And online apps like https://github.com/metinkale38/prayer-times-android

**Comment by AI-IMED** [comment]
> We can also add a "Custom" method which will allow users to enter their own angles

The feature of entering custom angles looks interesting, I'm eagerly awaiting the feature to try it, because I do not know my angle accurately, and to see if it will contribute to adjusting the accuracy of the calculations that I need.

**Comment by meypod** [comment]
methods other than Algeria and Unified London don't match exactly with their official websites when checked inside the five prayers app

**Comment by meypod** [comment]
Will add Algeria method as that's the only matching one

Unified London is for East London mosque and they use moonsighting committee method adjusted to their location so probably not going to add that
For morocco, tunisia and malaysiya will later research and try to find what parameters they use exactly

**Comment by meypod** [comment]
Thanks for the effort
but what we need is official calculation parameters/angles, the "how to" of their calculation
what they are providing is calculated times, not what we need

**Comment by meypod** [comment]
> You mean these websites can't be useful at all in knowing the angles?

yeah

I think something like this:

"What methods and parameters do you use to calculate the prayer times ? for example some apps that calculate adhan offline use Fajr, Isha and Maghrib angles. Can you provide the numbers or formulas that you use or a link to the official documents that explains what you use ?"

**Comment by meypod** [comment]
Bosnia source code didn't include parameters required for calculating, it only has times

also, the official websites are the source of truth for their times, if other apps are different from them, then the apps are incorrect

**Comment by meypod** [comment]
The website you contacted (ucide.org) does not seem to be related to the website you mentioned (hispanomuslim.es)

if they are related, you should have asked for method that is used to generate times on hispanomuslim.es, not in general.

wherever there's a time table, there's a method used to calculate it. we need the details of the method used to calculate it. anything else is not needed.

**Comment by meypod** [comment]
> They have also provided a website that explains how to calculate prayer times by yourself. I hope you like mathematics :)

I use adhan js which does all the mathmatics. I can't touch the internal calculations myself with my limited knowledge

I guess no need to add it if it's same as Muslim world league

**Comment by meypod** [comment]
Thanks for the time you have put into this @TorCzaf 
I am busy with other things at the moment but I will check them soon

**Comment by meypod** [comment]
kafiln/salati does not calculate times itself

**Comment by meypod** [comment]
oh i see you have asked and the owner already has said in the response

so yeah, I can't find anything on morocco either

**Comment by meypod** [comment]
For Malaysiya neither 18,18 nor 20,18 matches with https://www.e-solat.gov.my/index.php But 20,18 is a little closer

For Egypt 19.5,17.5 matches with Al-Azhar mosque time https://mawaqit.net/en/masjid-alazhar-le-caire-4293024-egypt , But 20,18 got like 2 minute difference, is there any official link for 20,18 angles?

France 18 and 15 seems more closer to Grand mosque of Paris prayer times but still doesn't match exactly, so will add France 18 and 15
